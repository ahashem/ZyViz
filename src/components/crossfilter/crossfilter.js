const crossfilter = require('crossfilter2');

export default crossfilter;
