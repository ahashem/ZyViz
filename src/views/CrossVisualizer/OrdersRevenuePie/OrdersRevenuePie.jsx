import React, { Component } from 'react';
import PropTypes from 'prop-types';
import PieChart from '../../../components/PieChart/PieChart';
import FlexGridRow from '../../../components/flex-wrapper/FlexGridRow';
import { Card } from 'antd';

/**
 * @component OrdersRevenuePie
 */
class OrdersRevenuePie extends Component {
  static propTypes = {
    orders: PropTypes.shape({}),
    dimensions: PropTypes.shape({}),
  };

  static defaultProps = {
    orders: {},
    dimensions: {},
  };

  /**
   * Receive a crossfilter dimensions, create groups to use for charts
   * @param {Object} dimensions
   * @return {Object<Array<crossfilter.Grouping>>}
   */
  prepareOrdersCountData = (dimensions) => {
    // Serialize and group data using crossfilter to be
    // for-example: {key: 'paymentMethod', value: 20}

    // Orders by paymentMethod
    const ordersByPaymentMethod = dimensions.ordersByPaymentMethod;
    const ordersByPaymentMethodData = ordersByPaymentMethod.group().reduceSum(order => order.orderAmount).all();

    // Orders by orderAmount
    // Less than $10, $10-$20, $20-40, $40-$70, $70 or more
    const ordersBySize = dimensions.ordersBySize;
    const ordersByAmountData = ordersBySize.group().reduceSum(order => order.orderAmount).all();

    // Orders by Order time
    // Morning 6am-12pm, Afternoon 12-5pm, Evening 5-8pm, Night 8pm-6am
    const ordersByTimeOfDay = dimensions.ordersByTimeOfDay;
    const ordersByTimeData = ordersByTimeOfDay.group().reduceSum(order => order.orderAmount).all();

    return {
      ordersByPaymentMethodData,
      ordersByAmountData,
      ordersByTimeData,
    };

  };

  /**
   * Uplift filter triggers
   * @param {*|string} dimension
   * @param {string} selectedKey
   * @param {Boolean} deselected
   */
  crossFilterSelected = (dimension, selectedKey, deselected = false) => {
    const { onFilter } = this.props;
    onFilter(dimension, selectedKey, deselected);
  };


  render() {
    const { dimensions, loading } = this.props;
    const ordersCrossed = dimensions ? this.prepareOrdersCountData(dimensions) : null;

    return (
      <div>
        <FlexGridRow justify="space-around" gutter={24}>

          <Card title="Revenue By Payment Method" bordered hoverable loading={loading}>
            <PieChart
              name='revenueByPaymentMethod'
              data={ordersCrossed && ordersCrossed.ordersByPaymentMethodData}
              onClick={(selectedIndex, deselected) => this.crossFilterSelected(
                'ordersByPaymentMethod',
                ordersCrossed.ordersByPaymentMethodData[`${selectedIndex}`].key,
                deselected
              )}
              x="key"
              y="value"
            />
          </Card>

          <Card title="Revenue By Order Size" bordered hoverable loading={loading}>
            <PieChart
              name='revenueBySize'
              data={ordersCrossed && ordersCrossed.ordersByAmountData}
              onClick={(selectedIndex, deselected) => this.crossFilterSelected(
                'ordersBySize',
                ordersCrossed.ordersByAmountData[`${selectedIndex}`].key,
                deselected
              )}
              x="key"
              y="value"
            />
          </Card>
          <Card title="Orders Revenue By Time of the Day" bordered hoverable loading={loading}>
            <PieChart
              name='revenueByTimeOfDay'
              data={ordersCrossed && ordersCrossed.ordersByTimeData}
              onClick={(selectedIndex, deselected) => this.crossFilterSelected(
                'ordersByTimeOfDay',
                ordersCrossed.ordersByTimeData[`${selectedIndex}`].key,
                deselected
              )}
              x="key"
              y="value"
            />
          </Card>

        </FlexGridRow>
      </div>
    );
  }
}

export default OrdersRevenuePie;
