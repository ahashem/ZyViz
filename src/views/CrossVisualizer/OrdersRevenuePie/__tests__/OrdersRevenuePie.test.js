import React from 'react';
import { shallow } from 'enzyme';

import OrdersRevenuePie from '../OrdersRevenuePie';
import crossfilter from '../../../../components/crossfilter/crossfilter';

import mockOrders from '../../../../utils/test-utils/mock-test-orders';

describe('<OrdersRevenuePie /> Component', () => {

  const mockOrdersCrossed = crossfilter(mockOrders);

  test('should render children and match snapshot', () => {

    const rendered = shallow(<OrdersRevenuePie orders={mockOrdersCrossed}/>);

    expect(rendered).toMatchSnapshot();
  });

});
