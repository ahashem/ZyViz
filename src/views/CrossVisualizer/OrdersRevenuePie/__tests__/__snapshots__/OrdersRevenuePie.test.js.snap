// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrdersRevenuePie /> Component should render children and match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <OrdersRevenuePie
    orders={
      Object {
        "add": [Function],
        "all": [Function],
        "allFiltered": [Function],
        "dimension": [Function],
        "groupAll": [Function],
        "isElementFiltered": [Function],
        "onChange": [Function],
        "remove": [Function],
        "size": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <FlexGridRow
        gutter={48}
        justify="space-around"
      >
        <PieChart
          data={null}
          onClick={[Function]}
          x="key"
          y="value"
        />
        <PieChart
          data={null}
          onClick={[Function]}
          x="key"
          y="value"
        />
        <PieChart
          data={null}
          onClick={[Function]}
          x="key"
          y="value"
        />
      </FlexGridRow>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <PieChart
            data={null}
            onClick={[Function]}
            x="key"
            y="value"
          />,
          <PieChart
            data={null}
            onClick={[Function]}
            x="key"
            y="value"
          />,
          <PieChart
            data={null}
            onClick={[Function]}
            x="key"
            y="value"
          />,
        ],
        "gutter": 48,
        "justify": "space-around",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "data": null,
            "onClick": [Function],
            "x": "key",
            "y": "value",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "data": null,
            "onClick": [Function],
            "x": "key",
            "y": "value",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "data": null,
            "onClick": [Function],
            "x": "key",
            "y": "value",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <FlexGridRow
          gutter={48}
          justify="space-around"
        >
          <PieChart
            data={null}
            onClick={[Function]}
            x="key"
            y="value"
          />
          <PieChart
            data={null}
            onClick={[Function]}
            x="key"
            y="value"
          />
          <PieChart
            data={null}
            onClick={[Function]}
            x="key"
            y="value"
          />
        </FlexGridRow>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <PieChart
              data={null}
              onClick={[Function]}
              x="key"
              y="value"
            />,
            <PieChart
              data={null}
              onClick={[Function]}
              x="key"
              y="value"
            />,
            <PieChart
              data={null}
              onClick={[Function]}
              x="key"
              y="value"
            />,
          ],
          "gutter": 48,
          "justify": "space-around",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "data": null,
              "onClick": [Function],
              "x": "key",
              "y": "value",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "data": null,
              "onClick": [Function],
              "x": "key",
              "y": "value",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "data": null,
              "onClick": [Function],
              "x": "key",
              "y": "value",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
